### PR TITLE
add a non-mutating incremental resolver interface

### PR DESCRIPTION
**Pull request by froydnj**
### Motivation

When we're running queries on stale `GlobalState` (#5407), we may need to run resolver in a way that modifies the affected trees we're looking at, but deliberately does not modify the `GlobalState` we're using.  (We expect that either a) the new `GlobalState` resulting from a successful ongoing typecheck will replace it or b) mutating it for temporary changes, but then having to put it back without those changes if the ongoing typecheck is cancelled for some reason, is fraught with problems.)

This PR attempts to provide such an interface with a minimum of code duplication, leaning on templates and `if constexpr` to magically wall off mutating sections of the resolver.  This way the compiler will automatically flag misuses of `const GlobalState` in the templated function(s) and it should be relatively obvious how to proceed to satisfy the compiler.

What's not clear to me is how effective this version of the resolver is going to be, since the WIP being filed here shuts off quite a bit of the resolver, including things like resolving superclasses, etc.  But maybe the expectation is that the mutating incremental resolver wouldn't have found that many changes to make anyway...?

### Test plan

Not sure how to test this.  Honestly this PR was just going to put the infrastructure in place; future PRs that actually call the non-mutating incremental resolver would need to figure out some sort of testing strategy.

### CODE REVIEW

File: resolver/resolver.cc [hunk 1]
```diff
@@ -3608,14 +3576,71 @@ class ResolveSignaturesWalk {
             }
             fillInInfoFromSig(ctx, overloadSym, sig.loc, sig.sig, isOverloaded, mdef);
         }
-        handleAbstractMethod(ctx, mdef);
+        // handleAbstractMethod called elsewhere
```
[the remark this answers is not in the document]
Comment (elliottt) [reply resolver/resolver.cc:3579]: I like moving `handleAbstractMethod` to `resolveSigs` 👍

File: resolver/resolver.cc [hunk 2]
```diff
@@ -2903,7 +2903,10 @@ class ResolveTypeMembersAndFieldsWalk {
         return tree;
     }
 
-    static vector<ast::ParsedFile> run(core::GlobalState &gs, vector<ast::ParsedFile> trees, WorkerPool &workers) {
+    template <typename StateType>
+    static vector<ast::ParsedFile> run(StateType &gs, vector<ast::ParsedFile> trees, WorkerPool &workers) {
+        static_assert(is_same_v<remove_const_t<StateType>, core::GlobalState>);
+        constexpr bool isConstStateType = is_const_v<StateType>;
```
[the remark this answers is not in the document]
Comment (froydnj) [reply resolver/resolver.cc:2909]: I support using `std::` wherever possible.  (Even so far as to never `using namespace std;`, but that is a hill to die on for a separate day.)

**Comment by froydnj** [comment]
I added a call to this in pipeline tests before we invoke the incremental resolver, and amazingly everything passes locally.

ptal @elliottt for the new test.

**Review by elliottt — APPROVED** [review]
I think the changes to the test runner make sense, though it looks like there's exactly one test failure for `test/testdata/resolver/infinite.rb`.

File: test/pipeline_test_runner.cc [hunk 3]
```diff
@@ -726,6 +726,19 @@ TEST_CASE("PerPhaseTest") { // NOLINT
...
+        // Non-mutating resolver.  Note that non-mutating resolver leaves GlobalState
+        // alone, but modifies trees with abandon, so we need to copy here.
+        vector<ast::ParsedFile> treesCopy;
+        for (auto &tree : trees) {
+            treesCopy.emplace_back(ast::ParsedFile{tree.tree.deepCopy(), tree.file});
+        }
+
+        move(resolver::Resolver::runIncrementalWithoutStateMutation(*gs, move(treesCopy)).result());
+        errorQueue->flushAllErrors(*gs);
+        errorCollector->drainErrors();
```
Comment (froydnj) [inline test/pipeline_test_runner.cc:739]: We used to have `ENFORCE(!gs->hadCriticalError());` here in an earlier commit.

However, this `ENFORCE` would fail on `test/testdata/resolver/infinite.rb`.  This test is testing recursive dealiasing of constants and deliberately sets out to make Sorbet hit its internal limit of recursive dealiasing.  As a result, we hit:

https://github.com/sorbet/sorbet/blob/ebe537d200780ff64dab06eaca250b6fda579221/core/Symbols.cc#L1935-L1939

and this error is defined as `StrictLevel::Internal`:

https://github.com/sorbet/sorbet/blob/ebe537d200780ff64dab06eaca250b6fda579221/core/errors/internal.h#L8

such errors are considered "critical" by `Error` and wind up setting a flag on the error queue.

I cribbed this `ENFORCE` from jez's analogous testing for non-mutating namer, and naively considered that it would enforce that we hadn't `ENFORCE`'d or somesuch.  In retrospect, this was a silly thing for me to assume: `ENFORCE`s should just fire without respect to the errors raised on `GlobalState` (they do not interact with Sorbet's error reporting mechanism), and a generic test runner shouldn't assume anything about whether a particular testcase does or does not trigger critical errors.  (The one test we have checking `hadCriticalError` uses a testcase specifically designed to trigger such an error.)

I think "critical" is the wrong terminology to use for this, but that is a separate issue.

Long story short, I have deleted the `ENFORCE(!gs->hadCriticalError());`, partly in an effort to make tests pass, but partly because I think it really is the right thing to do.  Ideally folks will agree with the rationale above.  If not, we can discuss it later.

File: resolver/resolver.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jez) [reply resolver/resolver.cc:3579]: ooc, why leave these comments behind? (we don't usually leave behind breadcrumbs documenting how sorbet has evolved over time)
